### DISCUSSION TITLE
Help generating XFMV Files

### DISCUSSION THREAD

**Discussion by Giulio64** [functionalresonance/FMV_Community_Edition#15] [Q&A] [answered]
Hi! I'm developing an application where I want to export a process as an XFMV files such that could be read by this application. I have some troubles understanding how to generate curves. I was able to narrow-down the following piece of code:

        public string curve
        {
            get
            {
                string[] _curve = new string[10]
                {
            drawTox.ToString("0.##", CultureInfo.InvariantCulture),
            drawToy.ToString("0.##", CultureInfo.InvariantCulture),
            drawFromx.ToString("0.##", CultureInfo.InvariantCulture),
            drawFromy.ToString("0.##", CultureInfo.InvariantCulture),
            drawAx.ToString("0.##", CultureInfo.InvariantCulture),
            drawAy.ToString("0.##", CultureInfo.InvariantCulture),
            drawBx.ToString("0.##", CultureInfo.InvariantCulture),
            drawBy.ToString("0.##", CultureInfo.InvariantCulture),
            drawIntx.ToString("0.##", CultureInfo.InvariantCulture),
            drawInty.ToString("0.##", CultureInfo.InvariantCulture)
                };
                return string.Join("|", _curve);
            }
        }
        public string curve2
        {
            get
            {
                string[] _curve2 = new string[13] {
            "M", drawFromx.ToString("0.##", CultureInfo.InvariantCulture), drawFromy.ToString("0.##", CultureInfo.InvariantCulture),
            "Q", drawAx.ToString("0.##", CultureInfo.InvariantCulture), drawAy.ToString("0.##", CultureInfo.InvariantCulture), drawIntx.ToString("0.##", CultureInfo.InvariantCulture), drawInty.ToString("0.##", CultureInfo.InvariantCulture),
            "Q", drawBx.ToString("0.##", CultureInfo.InvariantCulture), drawBy.ToString("0.##", CultureInfo.InvariantCulture), drawTox.ToString("0.##", CultureInfo.InvariantCulture), drawToy.ToString("0.##", CultureInfo.InvariantCulture)
            };
                return string.Join(" ", _curve2);
            }
        }

First of all, why do I need two curves? What's the maximum allowed sized of a project (In terms of X an Y dimensions)? What does that values represent? Thank you!

**Comment by zerprize** [comment, marked as answer]
Hi @Giulio64 
I can give some explanation but will start by pointing out that the curve data is not needed in an XFMV data file for it to be read by the application. You can refer to the documentation on the data format in the Wiki here:
https://github.com/functionalresonance/FMV_Community_Edition/wiki/Data-Format-and-.xfmv-file-structure

Note the section on Aspects, where the curve data is stored:
> The Aspects section was first used to store information specific to the display of a potential coupling (the line that connects two aspects from different Functions), such as the location of the label displaying the shared IDName. In later versions it also stores display settings and information on every coupling to speed up the graphical rendering and interaction of the model. Because this data is calculated and derived from the base data explained in the preceding chapters, it is not essential for a working data file and will be generated or overwritten by the FMV as needed.

If you are generating an XFMV file from another source, you should omit the Aspects section entirely, and the application will generate it as needed.

To explain the code you highlighted, both functions `curve` and `curve2` generate the same curve, but the first is for backward compatibility and simply stores the coordinates to draw the curve, while the second is used in the current application. If we then look at `curve2`, you will notice it uses three commands followed by coordinates, `M`, `Q`, and another `Q`. This is literally an SVG Path command. SVG graphics commands are used for all of the graphical rendering in the application. For more details on what `M` and `Q` do, search for SVG Path.

To answer the question about why there are two curves. Almost all coupling curves in the application are drawn using two quadratic Bézier curves. This gives the easiest way to draw a curve that changes direction but is always smooth. The point where the two curves join is the intermediate point where the coupling label is placed by default. The Bézier curve is also described by a control point. In the application, the control points of the two curves and the intermediate point where they join are always three points on a straight line so that the transition between the two curves is smooth, like joining two tangents.

And finally, there is no known limit to the x, y coordinates. If you find one, please let us know!

**Reply by Giulio64** [reply]
@zerprize Hi and thank you so much for the reply, it was so much helpful!

Best Regards,
Giulio
